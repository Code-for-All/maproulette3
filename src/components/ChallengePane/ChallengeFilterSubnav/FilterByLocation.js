import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { Link } from 'react-router-dom'
import _map from 'lodash/map'
import _isEmpty from 'lodash/isEmpty'
import { injectIntl, FormattedMessage } from 'react-intl'
import { ChallengeLocation,
         locationLabels }
       from '../../../services/Challenge/ChallengeLocation/ChallengeLocation'
import Dropdown from '../../Dropdown/Dropdown'
import ButtonFilter from './ButtonFilter'
import messages from './Messages'

/**
 * FilterByLocation displays a nav dropdown containing options for filtering
 * challenges by geographic location, such as Within Map Bounds and Near Me.
 * The redux store is updated to reflect the selected option.
 *
 * @author [Neil Rotstan](https://github.com/nrotstan)
 */
export class FilterByLocation extends Component {
  /**
   * Update the challenge filter with the selected value
   *
   * @private
   */
  updateFilter = value => {
    if (_isEmpty(value)) {
      this.props.removeSearchFilters(['location'])
    }
    else {
      // For nearMe, we actually use the withinMapBounds setting -- we just
      // also set the map bounds to be near the user.
      if (value === ChallengeLocation.nearMe) {
        // Note: repositioning the map will automatically trigger an update of the
        // bounded challenges, so we don't need to request an update here.
        this.props.setSearchFilters({location: ChallengeLocation.withinMapBounds})
        this.props.locateMapToUser(this.props.user)
      }
      else {
        this.props.setSearchFilters({location: value})
      }
    }
  }

  render() {
    const localizedLocationLabels = locationLabels(this.props.intl)

    const menuItems = _map(ChallengeLocation, (location, name) => (
      <li key={location}>
        <Link to={{}} onClick={() => this.updateFilter(location)}>
          {localizedLocationLabels[name]}
        </Link>
      </li>
    ))

    // Add 'Any' option to start of dropdown
    menuItems.unshift(
      <li key='any'>
        <Link to={{}} onClick={() => this.updateFilter(null)}>
          {localizedLocationLabels.any}
        </Link>
      </li>
    )

    return (
      <Dropdown
        className="mr-dropdown--flush xl:mr-border-l xl:mr-border-white-10 mr-p-6 mr-pl-0 xl:mr-pl-6"
        button={
          <ButtonFilter
            type={<FormattedMessage {...messages.locationLabel} />}
            selection={
              _isEmpty(this.props.searchFilters.location) ?
              localizedLocationLabels.any :
              localizedLocationLabels[this.props.searchFilters.location]
            }
          />
        }
      >
        <ol className="mr-list-dropdown mr-list-dropdown--ruled">
          {menuItems}
        </ol>
      </Dropdown>
    )
  }
}

FilterByLocation.propTypes = {
  /** The current map bounds */
  mapBounds: PropTypes.object,
  /** Invoked to update the challenge location filter */
  setSearchFilters: PropTypes.func.isRequired,
  /** Invoked to clear the challenge location filter */
  removeSearchFilters: PropTypes.func.isRequired,
  /** Invoked when the user chooses the 'near me' option */
  locateMapToUser: PropTypes.func.isRequired,
  /** The current value of the challenge filter */
  searchFilters: PropTypes.object,
  /** The current logged-in user, if any */
  user: PropTypes.object,
}

FilterByLocation.defaultProps = {
  searchFilters: {},
}

export default injectIntl(FilterByLocation)
