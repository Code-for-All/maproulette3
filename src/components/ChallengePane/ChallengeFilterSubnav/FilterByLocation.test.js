import React from 'react'
import { FilterByLocation } from './FilterByLocation'
import { cloneDeep as _cloneDeep } from 'lodash'

const propsFixture = {
  user: {
    id: 11,
    savedChallenges: [],
  },
  challenges: [
    {
      id: 309,
      name: "Challenge 309",
      blurb: "Challenge 309 blurb",
      description: "Challenge 309 description",
      parent: {
        displayName: "foo",
      }
    },
    {
      id: 311,
      name: "Challenge 311",
      blurb: "Challenge 311 blurb",
      description: "Challenge 311 description",
      parent: {
        displayName: "bar",
      }
    }
  ],
  challengeFilter: {
    difficulty: 1
  },
  mapBounds: {
    locator: {bounds: 1}
  },
}

let basicProps = null

beforeEach(() => {
  basicProps = _cloneDeep(propsFixture)
  basicProps.setChallengeFilters = jest.fn()
  basicProps.removeChallengeFilters = jest.fn()
  basicProps.locateMapToUser = jest.fn()
  basicProps.updateBoundedChallenges = jest.fn()
  basicProps.intl = {formatMessage: jest.fn()}
})

test("it renders with props as expected", () => {
  const wrapper = shallow(
    <FilterByLocation {...basicProps} />
  )

  expect(wrapper).toMatchSnapshot()
})

test("it calls setChallengeFilters/locateMapToUser if an onChange occurs with 'nearMe'", () => {
  const wrapper = mount(
    <FilterByLocation {...basicProps} />
  )

  wrapper.instance().updateFilter({value: 'nearMe'})
  expect(basicProps.setChallengeFilters).toBeCalledWith({"location": "withinMapBounds"})
  expect(basicProps.locateMapToUser).toBeCalledWith({"id": 11, "savedChallenges": []})
})


test("it calls setChallengeFilters/updateBoundedChallenges if an onChange occurs with withinMapBounds", () => {
  const wrapper = mount(
    <FilterByLocation {...basicProps} />
  )

  wrapper.instance().updateFilter({value: 'withinMapBounds'})
  expect(basicProps.setChallengeFilters).toBeCalledWith({"location": "withinMapBounds"})
  expect(basicProps.updateBoundedChallenges).toBeCalledWith(1)
})

test("it calls removeChallengeFilters if an onChange occurs with a null value", () => {
  const wrapper = mount(
    <FilterByLocation {...basicProps} />
  )

  wrapper.instance().updateFilter({value: null})
  expect(basicProps.removeChallengeFilters).toBeCalledWith(['location'])
})
