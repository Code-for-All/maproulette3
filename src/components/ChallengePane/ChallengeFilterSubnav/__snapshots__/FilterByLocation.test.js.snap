// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders with props as expected 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <FilterByLocation
    asMenuList={false}
    challenges={
        Array [
            Object {
              "blurb": "Challenge 309 blurb",
              "description": "Challenge 309 description",
              "id": 309,
              "name": "Challenge 309",
              "parent": Object {
                "displayName": "foo",
              },
            },
            Object {
              "blurb": "Challenge 311 blurb",
              "description": "Challenge 311 description",
              "id": 311,
              "name": "Challenge 311",
              "parent": Object {
                "displayName": "bar",
              },
            },
          ]
    }
    intl={
        Object {
            "formatMessage": [Function],
          }
    }
    locateMapToUser={[Function]}
    mapBounds={
        Object {
            "challenges": Object {
              "bounds": 1,
            },
          }
    }
    removeSearchFilters={[Function]}
    searchFilters={
        Object {
            "difficulty": 1,
          }
    }
    setSearchFilters={[Function]}
    user={
        Object {
            "id": 11,
            "savedChallenges": Array [],
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "eventTypes": Array [
        "mousedown",
        "touchstart",
      ],
      "excludeScrollbar": false,
      "label": <FormattedMessage
        defaultMessage="Location"
        id="ChallengeFilterSubnav.filter.location.label"
        values={Object {}}
/>,
      "onChange": [Function],
      "options": Array [
        Object {
          "key": "any",
          "text": "Anywhere",
          "value": undefined,
        },
        Object {
          "key": "withinMapBounds",
          "text": "Within Map Bounds",
          "value": "withinMapBounds",
        },
        Object {
          "key": "intersectingMapBounds",
          "text": "Intersecting Map Bounds",
          "value": "intersectingMapBounds",
        },
        Object {
          "key": "nearMe",
          "text": "Near Me",
          "value": "nearMe",
        },
      ],
      "outsideClickIgnoreClass": "ignore-react-onclickoutside",
      "placeholder": "Anywhere",
      "preventDefault": false,
      "stopPropagation": false,
      "value": undefined,
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "eventTypes": Array [
          "mousedown",
          "touchstart",
        ],
        "excludeScrollbar": false,
        "label": <FormattedMessage
          defaultMessage="Location"
          id="ChallengeFilterSubnav.filter.location.label"
          values={Object {}}
/>,
        "onChange": [Function],
        "options": Array [
          Object {
            "key": "any",
            "text": "Anywhere",
            "value": undefined,
          },
          Object {
            "key": "withinMapBounds",
            "text": "Within Map Bounds",
            "value": "withinMapBounds",
          },
          Object {
            "key": "intersectingMapBounds",
            "text": "Intersecting Map Bounds",
            "value": "intersectingMapBounds",
          },
          Object {
            "key": "nearMe",
            "text": "Near Me",
            "value": "nearMe",
          },
        ],
        "outsideClickIgnoreClass": "ignore-react-onclickoutside",
        "placeholder": "Anywhere",
        "preventDefault": false,
        "stopPropagation": false,
        "value": undefined,
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
